tests/unit/conftest.py

```diff
@@ -228,6 +228,39 @@ def harness_with_postgresql_fixture(
     return harness
 
 
+@pytest.fixture(name="harness_with_saml")
+def harness_with_saml_fixture(
```

Comment:
Sorry for the delay. This was a way to force harness reinitialize the charm was would happen in Juju since the charm we rely on charm_state to get information from the observers and this variable is updated only in the __init__.
I don't remember where I saw this way of forcing but I saw some charms doing this way like [here](https://github.com/canonical/prometheus-k8s-operator/blob/main/tests/unit/test_remote_write.py#L290).

Reply:
So I removed the `harness._framework` internals fiddling, and the test that failed was `test_enable_saml_success`, so I started editing that. A simpler and more "Harness-native" way to write that test would be to set up the relation data this function needs, then call `harness.begin()`. This avoids the re-init problem and avoids this complicated fixture altogether:

```diff
diff --git a/tests/unit/test_synapse_workload.py b/tests/unit/test_synapse_workload.py
index 816686e..e834e26 100644
--- a/tests/unit/test_synapse_workload.py
+++ b/tests/unit/test_synapse_workload.py
@@ -13,8 +13,9 @@ import pytest
 import yaml
 from ops.testing import Harness
 
+from charm import SynapseCharm
 import synapse
-from constants import SYNAPSE_CONFIG_PATH
+from constants import SYNAPSE_CONFIG_PATH, TEST_SERVER_NAME, SYNAPSE_CONTAINER_NAME
 
 
 def test_enable_metrics_success(monkeypatch: pytest.MonkeyPatch):
@@ -67,13 +68,17 @@ def test_enable_metrics_error(monkeypatch: pytest.MonkeyPatch):
         synapse.enable_metrics(container_mock)
 
 
-def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.MonkeyPatch):
-    """
-    arrange: set mock container with file.
-    act: change the configuration file.
-    assert: new configuration file is pushed and SAML is enabled.
-    """
-    harness = harness_with_saml
+def test_enable_saml_success(monkeypatch: pytest.MonkeyPatch):
+    # Arrange: set up harness and container filesystem
+    harness = Harness(SynapseCharm)
+    harness.update_config({"server_name": TEST_SERVER_NAME})
+    relation_id = harness.add_relation("saml", "saml-integrator")
+    harness.add_relation_unit(relation_id, "saml-integrator/0")
+    harness.update_relation_data(relation_id, "saml-integrator", {
+        "entity_id": "https://login.staging.ubuntu.com",
+        "metadata_url": "https://login.staging.ubuntu.com/saml/metadata",
+    })
+    harness.begin()
     config_content = """
     listeners:
         - type: http
@@ -89,16 +94,12 @@ def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.Mon
     monkeypatch.setattr(container_mock, "pull", pull_mock)
     monkeypatch.setattr(container_mock, "push", push_mock)
 
+    # Act: write the Synapse config file with SAML enabled
     synapse.enable_saml(container_mock, harness.charm._charm_state)
 
+    # Assert: ensure config file was written correctly
     assert pull_mock.call_args[0][0] == SYNAPSE_CONFIG_PATH
     assert push_mock.call_args[0][0] == SYNAPSE_CONFIG_PATH
-    saml_relation = harness.model.get_relation("saml")
-    assert saml_relation
-    saml_relation_data = harness.get_relation_data(saml_relation.id, "saml-integrator")
-    assert saml_relation_data
-    entity_id = saml_relation_data.get("entity_id")
-    metadata_url = saml_relation_data.get("metadata_url")
     expected_config_content = {
         "listeners": [
             {"type": "http", "x_forwarded": True, "port": 8080, "bind_addresses": ["::"]}
@@ -106,9 +107,9 @@ def test_enable_saml_success(harness_with_saml: Harness, monkeypatch: pytest.Mon
         "saml2_enabled": True,
         "saml2_config": {
             "sp_config": {
-                "metadata": {"remote": [{"url": metadata_url}]},
+                "metadata": {"remote": [{"url": "https://login.staging.ubuntu.com/saml/metadata"}]},
                 "service": {"sp": {"entityId": "https://server-name-configured.synapse.com"}},
-                "name": [entity_id, "en"],
+                "name": ["https://login.staging.ubuntu.com", "en"],
             },
             "user_mapping_provider": {
                 "config": {"mxid_source_attribute": "username", "mxid_mapping": "dotreplace"}
```

[Full code for modified test here.](https://pastebin.canonical.com/p/nHZYt7YBr4/). Still, I found the number of fixture layers here very hard to follow. To understand the setup for this function, we have to go through 3 layers of fixtures:

Side note: not at all related to this PR, but I find pytest's magical match-by-name fixtures hard to follow, especially when they're in a magically-imported file (conftest.py). At least we have ripgrep! :-)

```
test_enable_saml_success, which calls
harness_with_saml, which calls
harness_server_name_configured, which calls
harness_fixture
```

In my experience it's often better to cut through all that and just set up what you need in each test, like I've done above. Even if it's a bit more repetitive, it makes individual tests much easier to understand and modify without breaking other tests. Or you can use very targetted fixtures that set up a specific thing and "composing" them, rather than "inheriting" them like we did here.

Note also how in the modified version I've hard-coded the expected entity_id and metadata_url. I think this simplified and also avoids probing into the relation data (an implementation detail as far as this function is concerned).

One more thing. Now that we have @weiiwang01's excellent `Harness.get_filesystem_root`, I'd simplify further and avoid the mocking of push and pull. [Example of the test with get_filesystem_root here.](https://pastebin.canonical.com/p/229QXwpxS4/)

Reply:
Thank you for taking the time to write all these examples! I updated the test_enable_saml_success and the others will be changed in a follow-up PR.